### PR TITLE
[C/PyTorch] Add max_t support for THD

### PR DESCRIPTION
**Pull request by cyanguwa**
# Description

This PR takes advantage of cuDNN 9.6+ and reduces the memory requirement for intermediate tensors and workspace tensors, for `qkv_format=thd`.

## Type of change

- [ ] Documentation change (change only to the documentation, either a fix or a new content)
- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Infra/Build change
- [ ] Code refractor

## Changes

Please list the changes introduced in this PR:

- Add `max_t` support for THD attention

# Checklist:

- [x] I have read and followed the [contributing guidelines](https://github.com/NVIDIA/TransformerEngine/blob/main/CONTRIBUTING.rst)
- [x] The functionality is complete
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 1]
```diff
@@ -321,7 +345,7 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
     const size_t actual_seqlen_workspace_size = 2 * num_bytes_per_seqlen;
     const size_t num_bytes_per_ragged_offset =
         alignTo<16>((b + 1) * typeToSize(ragged_offset_type));
-    const size_t seqlen_offsets_workspace_size = 4 * num_bytes_per_ragged_offset;
+    const size_t seqlen_offsets_workspace_size = 5 * num_bytes_per_ragged_offset;
```
[the remark this answers is not in the document]
Comment (mgoldfarb-nvidia) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:348]: To be pedantic do we need to also check for cudnn version >= 9.6 when we use the additional offset?

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 2]
```diff
@@ -276,10 +285,24 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
...
+        offset_stats =
+            mha_graph->tensor(fe::graph::Tensor_attributes()
+                                  .set_name("offset_stats")
+                                  .set_dim({b + 1, 1, 1, 1})
+                                  .set_stride({1, 1, 1, 1})
+                                  .set_data_type(get_cudnn_fe_dtype(ragged_offset_type)));
+        Stats->set_output(true)
+            .set_data_type(fe::DataType_t::FLOAT)
+            .set_dim({b, h, s_q, 1})
+            .set_stride({h * s_q, 1, h, 1})
```
Comment (mgoldfarb-nvidia) [inline transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:298]: Is this stride correct? How come it changes here?

Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu:298]: Yes, we are switching from `{b, h, s, 1}` to `{b, t, h, 1}` where `t` is the `max_t` and the stride for `b` is ignored.

**Review by mgoldfarb-nvidia — APPROVED** [review]

File: transformer_engine/common/fused_attn/utils.h [hunk 3]
```diff
@@ -11,6 +11,7 @@
 #include <cudnn_frontend.h>
 #include <cudnn_frontend_utils.h>
 
+#include <cmath>
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/utils.h]: Moved to `utils.cu`.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu (hunk 1 above)
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Added more checks for both `is_ragged` and cuDNN versions.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 4]
```diff
@@ -73,8 +73,16 @@ void fused_attn_arbitrary_seqlen_fwd_impl(
                      (mask_type == NVTE_Mask_Type::NVTE_PADDING_CAUSAL_MASK));
   bool is_dropout = (is_training && dropout_probability != 0.0f);
   bool is_ragged = (nvte_get_qkv_format(layout) == NVTE_QKV_Format::NVTE_THD);
+
+  // keep original batch size because cu_seqlens are created with [b+1] shape
+  int64_t orig_b = b;
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Renamed `orig_b` to `actual_b`.

File: transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu [hunk 5]
```diff
@@ -414,6 +441,15 @@ void fused_attn_arbitrary_seqlen_bwd_impl(
   const auto cudnn_runtime_version = cudnnGetVersion();
   const int device_id = cuda::current_device();
   const int sm_arch_ = cuda::sm_arch(device_id);
+  // keep original batch size because cu_seqlens are created with [b+1] shape
+  int64_t orig_b = b;
+  if (is_ragged) {
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/fused_attn_f16_arbitrary_seqlen.cu]: Done.

File: transformer_engine/common/fused_attn/utils.cu [hunk 6]
```diff
@@ -393,26 +397,50 @@ __device__ void cu_seqlens_padded_to_offsets_impl(NVTE_QKV_Layout_Group layout_g
         offsets_v[tid] = offsets_k[tid];
         break;
     }
+  } else if (tid < max_b + 1) {
```
[the remark this answers is not in the document]
Comment (cyanguwa) [reply transformer_engine/common/fused_attn/utils.cu]: Good point.

**Comment by cyanguwa** [comment]
Test cuDNN 9.5.1: 19711074

**Pull request merged**